Linear dbt bigframes blog (#7625)
fishtown-analytics/docs.getdbt.com: website/blog/2025-07-111-train-linear-regression-model-dbt-bigframes.md

---
title: "How to train a linear regression model with dbt and BigFrames"
description: "How to build a scalable linear regression model by combining dbt's modular orchestration with BigFrames' in-database Python execution in BigQuery."
slug: train-linear-dbt-bigframes
authors: [jialuo_chen]

tags: [analytics craft]
hide_table_of_contents: false

date: 2025-07-11
is_featured: true
---

## Introduction to dbt and BigFrames

> **dbt**: A framework for transforming data in modern data warehouses using modular SQL or Python. It enables data teams to develop analytics code collaboratively and efficiently by applying software engineering best practices such as version control, modularity, portability, CI/CD, testing, and documentation. For more information, refer to [What is dbt?](/docs/introduction#dbt).

Author: ```suggestion
**dbt**: A framework for transforming data in modern data warehouses using modular SQL or Python. dbt enables data teams to develop analytics code collaboratively and efficiently by applying software engineering best practices such as version control, modularity, portability, CI/CD, testing, and documentation. For more information, refer to [What is dbt?](/docs/introduction#dbt).
```

> **dbt**: A framework for transforming data in modern data warehouses using modular SQL or Python. dbt enables data teams to develop analytics code collaboratively and efficiently by applying software engineering best practices such as version control, modularity, portability, CI/CD, testing, and documentation. For more information, refer to [What is dbt?](/docs/introduction#dbt).

Author: ```suggestion
**dbt**: A framework for transforming data in modern data warehouses using modular SQL or Python. dbt enables data teams to develop analytics code collaboratively and efficiently by applying software engineering best practices such as version control, modularity, portability, CI/CD, testing, and documentation. For more information, refer to [What is dbt?](/docs/introduction#dbt)
```

> **BigQuery DataFrames (BigFrames)**: An open-source Python library offered by Google. BigFrames scales Python data processing by transpiling common Python data science APIs (Pandas and Scikit-learn) to BigQuery SQL.  

Reviewer: I believe both their brandings are lower case like dbt
```suggestion
**BigQuery DataFrames (BigFrames)**: An open-source Python library offered by Google. BigFrames scales Python data processing by transpiling common Python data science APIs (pandas and scikit-learn) to BigQuery SQL.  
```

> You can read more in the [official BigFrames guide](https://cloud.google.com/bigquery/docs/bigframes-intro) and view the [public BigFrames GitHub repository](https://github.com/GoogleCloudPlatform/bigframes).

Author: These links are coming up with 404s

Reviewer 2: The links from the google doc is correct. Please see below:

[official BigFrames guide](https://cloud.google.com/bigquery/docs/bigquery-dataframes-introduction)

[public BigFrames GitHub repository](https://github.com/googleapis/python-bigquery-dataframes)

> You can read more in the [official BigFrames guide](https://cloud.google.com/bigquery/docs/bigframes-intro) and view the [public BigFrames GitHub repository](https://github.com/GoogleCloudPlatform/bigframes).

Author: ```suggestion
You can read more in the [official BigFrames guide](https://cloud.google.com/bigquery/docs/bigquery-dataframes-introduction) and view the [public BigFrames GitHub repository](https://github.com/googleapis/python-bigquery-dataframes).
```

> By combining dbt with BigFrames—via the **dbt-bigquery adapter** (referred to as _"dbt-BigFrames"_)—you gain:

Reviewer: Wondering if we want to stick to our styling around adapter names here
```suggestion
By combining dbt with BigFrames via the `dbt-bigquery` adapter (referred to as _"dbt-BigFrames"_), you gain:
```

> - dbt’s modular SQL and Python modeling, dependency management with dbt.ref(), environment configurations, and data testing. With dbt Cloud, you also get job scheduling and monitoring.

Reviewer: New branding for dbt Cloud
```suggestion
- dbt’s modular SQL and Python modeling, dependency management with dbt.ref(), environment configurations, and data testing. With the cloud-based dbt platform, you also get job scheduling and monitoring.
```

- BigFrames’ ability to execute complex Python transformations (including machine learning) directly in BigQuery.

> `dbt-BigFrames` utilizes the **Colab Enterprise Notebook Executor Service** in a GCP project to run Python models. These notebooks execute BigFrames code, which is translated into BigQuery SQL.

Reviewer: The only instance I can find of this whole phrase in the Google docs brands it like this ([reference](https://cloud.google.com/bigquery/docs/dataframes-dbt))
```suggestion
`dbt-BigFrames` utilizes the **Colab Enterprise notebook executor service** in a GCP project to run Python models. These notebooks execute BigFrames code, which is translated into BigQuery SQL.
```

<!-- truncate -->

> Refer to the [BigFrames guide](https://cloud.google.com/bigquery/docs/bigframes-intro) or [dbt guides](/guides) to learn more.

To illustrate the practical impact of combining dbt with BigFrames, the following sections explore how this integration can streamline and scale a common machine learning task: training a linear regression model on large datasets.

## The power of dbt-BigFrames for large-scale linear regression

Linear regression is a cornerstone of predictive analytics, used in:

- Sales forecasting  
- Financial modeling  
- Demand planning  
- Real estate valuation

These tasks often require processing datasets too large for traditional in-memory Python. BigFrames alone solves this, but combining it with dbt offers a structured, maintainable, and production-ready way to train models or generate batch predictions on large data.

> ## “dbt-BigFrames” with ML: A Practical Example

Reviewer: For consistent casing amongst the headers
```suggestion
## “dbt-BigFrames” with ML: A practical example
```

We’ll walk through training a linear regression model using a **dbt Python model powered by BigFrames**, focusing on the structure and orchestration provided by dbt.

We’ll use the `epa_historical_air_quality` dataset from BigQuery Public Data (courtesy of the U.S. Environmental Protection Agency).

### Problem statement

Develop a machine learning model to predict atmospheric ozone levels using historical air quality and environmental sensor data, enabling more accurate monitoring and forecasting of air pollution trends.

**Key stages:**

1. **Data Foundation**: Transform raw source tables into an analysis-ready dataset.
2. **Machine learning Analysis**: Train a linear regression model on the cleaned data.

## Setting up your dbt project for BigFrames

### Prerequisites

- A Google Cloud account  

> - A dbt Platform or Core setup  

Reviewer: Branding for "platform" is lowercase
```suggestion
- A dbt platform or Core setup  
```

- Basic to intermediate SQL and Python  
- Familiarity with dbt using [Beginner dbt guides](/guides?level=Beginner)

### Sample `profiles.yml` for BigFrames

```yaml
my_epa_project:
  outputs:
    dev:
      compute_region: us-central1
      dataset: your_bq_dataset
      gcs_bucket: your_gcs_bucket
      location: US
      method: oauth
      priority: interactive
      project: your_gcp_project
      threads: 1
      type: bigquery
  target: dev
```

### Sample `dbt_project.yml`

```yaml
name: 'my_epa_project'
version: '1.0.0'
config-version: 2

models:
  my_epa_project:
    submission_method: bigframes
    notebook_template_id: 701881164074529xxxx  # Optional
    timeout: 6000
    example:
      +materialized: view
```

> ## The dbt Python Models for Linear Regression

Reviewer: Keeping the casing consistent with the other headers in this article
```suggestion
## The dbt Python models for linear regression
```

This project uses **two modular dbt Python models**:

> 1. `prepare_table.py` – Ingests and prepares data

Reviewer: ```suggestion
1. `prepare_table.py` &mdash; Ingests and prepares data
```

> 2. `prediction.py` – Trains the model and generates predictions

Reviewer: ```suggestion
2. `prediction.py` &mdash; Trains the model and generates predictions
```

> ### Part 1: Preparing the Table (`prepare_table.py`)

Reviewer: ```suggestion
### Part 1: Preparing the table (`prepare_table.py`)
```

    value_column = "sample_measurement"
    params_dfs = []

    table_param_dict = {
        "co_hourly_summary": "co",
        "no2_hourly_summary": "no2",
        "o3_hourly_summary": "o3",
        "pressure_hourly_summary": "pressure",
        "so2_hourly_summary": "so2",
        "temperature_hourly_summary": "temperature",
    }

    for table, param in table_param_dict.items():
        param_df = bpd.read_gbq(f"{dataset}.{table}", columns=index_columns + [value_column])
        param_df = param_df.sort_values(index_columns).drop_duplicates(index_columns).set_index(index_columns).rename(columns={value_column: param})
        params_dfs.append(param_df)

    wind_table = f"{dataset}.wind_hourly_summary"
    wind_speed_df = bpd.read_gbq(
        wind_table,
        columns=index_columns + [value_column],
        filters=[(param_column, "==", "Wind Speed - Resultant")]
    )
    wind_speed_df = wind_speed_df.sort_values(index_columns).drop_duplicates(index_columns).set_index(index_columns).rename(columns={value_column: "wind_speed"})
    params_dfs.append(wind_speed_df)

    df = bpd.concat(params_dfs, axis=1, join="inner").cache()
    return df.reset_index()
```

> ### Part 2: Training the Model and Making Predictions (`prediction.py`)

Reviewer: ```suggestion
### Part 2: Training the model and making predictions (`prediction.py`)
```

```python
def model(dbt, session):
    dbt.config(submission_method="bigframes", timeout=6000)

    df = dbt.ref("prepare_table")

    train_data_filter = (df.date_local.dt.year < 2017)
    test_data_filter = (df.date_local.dt.year >= 2017) & (df.date_local.dt.year < 2020)
    predict_data_filter = (df.date_local.dt.year >= 2020)

    index_columns = ["state_name", "county_name", "site_num", "date_local", "time_local"]
    df_train = df[train_data_filter].set_index(index_columns)
    df_test = df[test_data_filter].set_index(index_columns)
    df_predict = df[predict_data_filter].set_index(index_columns)

    X_train, y_train = df_train.drop(columns="o3"), df_train["o3"]
    X_test, y_test = df_test.drop(columns="o3"), df_test["o3"]
    X_predict = df_predict.drop(columns="o3")

    from bigframes.ml.linear_model import LinearRegression
    model = LinearRegression()
    model.fit(X_train, y_train)
    df_pred = model.predict(X_predict)

    return df_pred
```

> ## Running Your dbt ML Pipeline

Reviewer: ```suggestion
## Running your dbt ML pipeline
```

```bash
# Run all models
dbt run

# Or run just your new models
dbt run --select prepare_table prediction
```

> ## Key advantages of dbt + BigFrames for ML

Reviewer: The + symbol complicates the header link a bit, might be better to stick with a word here
```suggestion
## Key advantages of dbt and BigFrames for ML
```

- **Scalability & Efficiency**: Handle large datasets in BigQuery via BigFrames  
- **Simplified Workflow**: Use familiar APIs like `pandas` and `scikit-learn`  
- **dbt Orchestration**:
  - Dependency management with `dbt.ref()` and `dbt.source()`
  - Scheduled retraining with `dbt run`
  - Testing, documentation, and reproducibility

## Conclusion and next steps

By integrating **BigFrames** into your **dbt workflows**, you can build scalable, maintainable, and production-ready machine learning pipelines. While this example used linear regression, the same principles apply across other ML use cases with `bigframes.ml`.

> ## Feedback & support

Reviewer 2: I believe you’ve updated this section, but the links don’t appear to be working. Could you also double-check the other links to ensure everything is functioning correctly?

> ## Feedback & support

Reviewer: Same deal as above with the & symbol
```suggestion
## Feedback and support
```

- 📚 [dbt Support](/docs/dbt-support)  

> - 📨 Email feedback on BigFrames: `bigframes-feedback@google.com`  

Author: ```suggestion
- 📨 Email feedback on BigFrames: `bigframes-feedback@google.com`  
```

> - 📨 Email feedback on BigFrames: `bigframes-feedback@google.com`  

Reviewer: We can make this a hyperlink to help
```suggestion
- 📨 Email feedback on BigFrames: [bigframes-feedback@google.com](mailto:bigframes-feedback@google.com)
```

> - 🛠 [File issues on GitHub](https://github.com/GoogleCloudPlatform/bigframes)  

Author: ```suggestion
- 🛠 [File issues on GitHub](https://github.com/googleapis/python-bigquery-dataframes)  
```

> - 📬 [Subscribe to BigFrames updates](https://groups.google.com/g/bigframes-announce)

Author: ```suggestion
- 📬 [Subscribe to BigFrames updates](https://docs.google.com/forms/d/10EnDyYdYUW9HvelHYuBRC8L3GdGVl3rX0aroinbRZyc/viewform?edit_requested=true)
```

---
title: "How to train a linear regression model with dbt and BigFrames"
description: "How to build a scalable linear regression model by combining dbt's modular orchestration with BigFrames' in-database Python execution in BigQuery."
slug: train-linear-dbt-bigframes
authors: [jialuo_chen]

tags: [analytics craft]
hide_table_of_contents: false

date: 2025-07-11
is_featured: true
---

## Introduction to dbt and BigFrames

**dbt**: A framework for transforming data in modern data warehouses using modular SQL or Python. It enables data teams to develop analytics code collaboratively and efficiently by applying software engineering best practices such as version control, modularity, portability, CI/CD, testing, and documentation. For more information, refer to [What is dbt?](/docs/introduction#dbt).

> 

Author: ```suggestion
```

---

> title: "How to train a linear regression model with dbt and BigFrames?"

Author: ```suggestion
title: "How to train a linear regression model with dbt and BigFrames"
```

> title: "How to train  a Linear regression model with dbt and BigFrames?"

Author: ```suggestion
title: "How to train a linear regression model with dbt and BigFrames?"
```

description: "How to build a scalable linear regression model by combining dbt's modular orchestration with BigFrames' in-database Python execution in BigQuery."
slug: train-linear-dbt-bigframes
authors: [jialuo_chen]

tags: [analytics craft]
hide_table_of_contents: false

date: 2025-07-11
is_featured: true
---

## Introduction to dbt and BigFrames

**dbt**: A framework for transforming data in modern data warehouses using modular SQL or Python. It enables data teams to develop analytics code collaboratively and efficiently by applying software engineering best practices such as version control, modularity, portability, CI/CD, testing, and documentation. For more information, refer to [What is dbt?](/docs/introduction#dbt).

> > This blog post uses dbt Core (open-source, command-line interface), but you can also use dbt platform with minor configuration changes.

Reviewer 3: What changes? The model code will be identical, the only thing that is different is that platform users don't need to configure their own profiles.yml file. But the prereq's say "a dbt platform or Core setup" anyway. 
```suggestion
```

Author: Good shout, @joellabes !  I could just say you don't need to configure profiles.yml

I was mostly making sure people realize this isn't a core only thing.

**BigQuery DataFrames (BigFrames)**: An open-source Python library offered by Google. BigFrames scales Python data processing by transpiling common Python data science APIs (Pandas and Scikit-learn) to BigQuery SQL.  

You can read more in the [official BigFrames guide](https://cloud.google.com/bigquery/docs/bigframes-intro) and view the [public BigFrames GitHub repository](https://github.com/GoogleCloudPlatform/bigframes).

By combining dbt with BigFrames—via the **dbt-bigquery adapter** (referred to as _"dbt-BigFrames"_)—you gain:

- dbt’s modular SQL and Python modeling, dependency management with dbt.ref(), environment configurations, and data testing. With dbt Cloud, you also get job scheduling and monitoring.
- BigFrames’ ability to execute complex Python transformations (including machine learning) directly in BigQuery.

`dbt-BigFrames` utilizes the **Colab Enterprise Notebook Executor Service** in a GCP project to run Python models. These notebooks execute BigFrames code, which is translated into BigQuery SQL.

> <!-- truncate -->

Reviewer 4: so stoked we never have to worry about this not being here again!

Author: @dataders I totally smiled big when I got the error message!

    return df.reset_index()
```

### Part 2: Training the Model and Making Predictions (`prediction.py`)

```python
def model(dbt, session):
    dbt.config(submission_method="bigframes", timeout=6000)

    df = dbt.ref("prepare_table")

    train_data_filter = (df.date_local.dt.year < 2017)
    test_data_filter = (df.date_local.dt.year >= 2017) & (df.date_local.dt.year < 2020)
    predict_data_filter = (df.date_local.dt.year >= 2020)

    index_columns = ["state_name", "county_name", "site_num", "date_local", "time_local"]
    df_train = df[train_data_filter].set_index(index_columns)
    df_test = df[test_data_filter].set_index(index_columns)
    df_predict = df[predict_data_filter].set_index(index_columns)

    X_train, y_train = df_train.drop(columns="o3"), df_train["o3"]
    X_test, y_test = df_test.drop(columns="o3"), df_test["o3"]
    X_predict = df_predict.drop(columns="o3")

    from bigframes.ml.linear_model import LinearRegression
    model = LinearRegression()
    model.fit(X_train, y_train)
    df_pred = model.predict(X_predict)

    return df_pred
```

## Running Your dbt ML Pipeline

```bash
# Run all models
dbt run

# Or run in order

> dbt run --select prepare_table prediction

Reviewer 3: dbt always runs in DAG order, `--select` narrows down which models are built
```suggestion
# Or run just your new models
dbt run --select prepare_table prediction
```

## Introduction to dbt and BigFrames

**dbt**: A framework for transforming data in modern data warehouses using modular SQL or Python. It enables data teams to develop analytics code collaboratively and efficiently by applying software engineering best practices such as version control, modularity, portability, CI/CD, testing, and documentation. For more information, refer to [What is dbt?](/docs/introduction#dbt).

> > This blog post uses dbt Core (open-source, command-line interface), but you can also use dbt platform with minor configuration changes.

Author: ```suggestion
> This blog post uses dbt Core (open-source, command-line interface), but you can also use dbt platform which does not require configuring your own profiles.yml.
```
